### PR TITLE
Replace per-PVC leader election with per-cluster

### PR DESCRIPTION
**Pull request by wongma7**
continuing work in https://github.com/kubernetes-incubator/external-storage/pull/837

### CODE REVIEW

**Comment by wongma7** [comment]
@orainxiong please review. 1st commit is taken from your work at https://github.com/kubernetes-incubator/external-storage/pull/837 (thanks) 2nd commit is the replacement leader election. It will create an endpoints object in namespace kube-system with name equal to `provisionerName`, i.e. the value of storage class `provisioner` that the provisioner should watch. e.g. if the provisioner on the storageclass is `example.com/nfs` then the endpoint will be `kube-system/example.com-nfs`

Some open questions:
* should the namespace for the endpoints object be configurable instead of just `kube-system`? IMO it's okay to hardcode it as kube-system.
* is it bad to give provisioners endpoints write access? (don't think this is too big a deal, we already distribute clusterroles where provisioners have SECRETS access and it's up to users to audit it)
    * TODO for me: fix all the clusterrole.yamls throughout the repo (while at it, merge serviceaccount+clusterrole+clusterrolebinding yamls) so that they include this new permission


/cc @cofyc

**Comment by wongma7** [comment]
also TODO for me tomorrow: fix NFS docs, I think they reference hostpath daemonsets and such somewhere which won't be possible anymore :p

**Comment by orainxiong** [comment]
@wongma7 

LGTM! 

You have a better workaround, and I have closed related issues in `external-provisioner` repos.

If possible I will measure the effects of this pr for performance improvement with the same test case of `provisioning 100 pvc at once`.

**Comment by wongma7** [comment]
I don't really understand the bug, how common is it? How can tryAcquireOrRenew get stuck?  I don't want to bump all the k8s.io/* dependencies to master. :/ Since upstream did not think it's important enough to cherry-pick I am inclined to live with the bug for now.

File: lib/controller/controller.go [hunk 1]
```diff
@@ -141,10 +142,6 @@ type ProvisionController struct {
 	// when multiple controllers are running: they race to lock (lead) every PVC
 	// so that only one calls Provision for it (saving API calls, CPU cycles...)
```
Comment (cofyc) [inline lib/controller/controller.go]: this comment should be updated

File: lib/controller/controller.go [hunk 2]
```diff
@@ -1004,9 +952,13 @@ func (ctrl *ProvisionController) provisionClaimOperation(claim *v1.PersistentVol
 	//  A previous doProvisionClaim may just have finished while we were waiting for
 	//  the locks. Check that PV (with deterministic name) hasn't been provisioned
 	//  yet.
+	namespace := claim.GetNamespace()
 	pvName := ctrl.getProvisionedVolumeNameForClaim(claim)
-	volume, err := ctrl.client.CoreV1().PersistentVolumes().Get(pvName, metav1.GetOptions{})
-	if err == nil && volume != nil {
+	_, exists, err := ctrl.volumes.GetByKey(fmt.Sprintf("%s/%s", namespace, pvName))
```
Comment (cofyc) [inline lib/controller/controller.go]: If we check volume existence in `cache.Store`, perhaps we should wait informers are fully synced before running any controller logic.

File: lib/controller/controller.go [hunk 3]
```diff
@@ -613,46 +608,82 @@ func (ctrl *ProvisionController) forgetWork(queue workqueue.RateLimitingInterfac
...
+				if err != nil {
+					glog.Errorf("Failed to listen on %s: %v", address, err)
+				}
+			}, 5*time.Second)
+		}
+
+		go ctrl.claimController.Run(stopCh)
+		go ctrl.volumeController.Run(stopCh)
+		go ctrl.classController.Run(stopCh)
+
```
[the remark this answers is not in the document]
Comment (cofyc) [reply lib/controller/controller.go]: oh , should use `controller.HasSynced`, ctrl.xxxInformer is optional.
`cache.WaitForCacheSync(stopCh, ctrl.claimController.HasSynced, ctrl.volumeController.HasSynced, ctrl.classController.HasSynced)`

[the remark this answers is not in the document]
Comment (wongma7) [reply lib/controller/controller.go]: Good idea, thank you

While I was looking at this code, I also think it is a bug that we call `Run` for SharedInformers. Users of lib should be able to `Run` the SharedInformers whenever they want... May as well fix it while we are here IMO

**Comment by wongma7** [comment]
Last call for review @cofyc 

I will merge this tomorrow and tag a release. I plan to fix the nfs e2e tests in a separate pr. I’ve revamped them so they’re not so fragile

Other than the e2e testing I’ve done some local testing. Probably not sufficient but it can’t be helped. Anyway the code is identical to controller manager so we should be okay

**Comment by wongma7** [comment]
/lgtm
rebased for controller_test.go

**Comment by weherdh** [comment]
@wongma7 Hi, this change break Openshift testing... Openshift does not have resource `PodSecurityPolicy`, could you please also add the deployment files for Openshfit? And also provide the instruction how we deploy nfs-provisioner on Openshift. If it is needed, I'd like to open a bug for this. Thanks

**Comment by wongma7** [comment]
@funky81 please try the new RBAC with the latest release v3.0.0-k8s1.11

@weherdh sorry, I removed the SCC without a replacement https://github.com/kubernetes-incubator/external-storage/pull/892/files#diff-fbc2b7e3391a05df13aa2ae2e9e9831a. Please feel free to open a bug so I can track it. In openshift instead of creating PSP we create SCC and it should work right?

**Pull request merged**